### ISSUE TITLE
Optimization to deduplicate literals within dask graphs

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1795]
We would like to sometimes rewrite graphs like the following:

```python
x = np.ones(1000000)

dsk = {'a': (add, x, 1),
       'b': (sub, x, 1)}
```

to

```python
dsk = {'x': x,
      'a': (add, 'x', 1),
      'b': (sub, 'x', 1)}
```

This is a bit like the opposite operation of inlining.  Note that we pulled out the literal `x`, but not `1`.  We might want to pull out objects of a particular type or that match a particular predicate, such as being expensive to serialize.

cc @eriknw , in case you're interested.

**Comment by eriknw** [comment]
I should be able to get to this this week.

Operating in the domain of values is more complicated and is something we do very little of in `dask.core`.  How should we determine whether literals match?  Can we rely on `is` or `==` relationship?  Should this operation work for mutable and immutable types?  For `pandas` and `numpy` objects, do we need to do more stringent equivalence tests that includes comparing dtypes, and, if so, do these utility functions exist somewhere?

**Comment by mrocklin** [comment]
I would expect that we would deduplicate based on object equality, i.e. `is`.  One solution would be to do this in two passes.  In one we collect all literals, filter by the predicate, and then count frequencies.  In the second pass we replace duplicate literals with a key value.

**Comment by eriknw** [comment]
I call `is` identity and `==` equality comparison.  What about mutable objects?  If we can't easily use values as keys in a `dict`, things become more difficult.  If we rely on `is`, though, we can use `id(val) // 8` as a partial hash to keep the algorithmic complexity down.

**Comment by pitrou** [comment]
I think you certainly want to lookup by identity, both for mutable and immutable data.

**Comment by jcrist** [comment]
This was fixed by #1818, but was no longer deemed necessary. Closing for now, can be reopened (along with that PR) if ever needed.
